balancer/rls/internal/adaptive/adaptive.go

```diff
@@ -0,0 +1,143 @@
+// support for the same will be added.
+// * Duration: amount of recent history that will be taken into account for
+//   making client-side throttling decisions. A default of 30 seconds is used.
+// * Bins: number of bins to be used for bucketing historical data. A default
+//   of 100 is used.
+// * RatioForAccepts: ratio by which accepts are multiplied, typically a value
+//   slightly larger than 1.0. This is used to make the throttler behave as if
+//   the backend had accepted more requests than it actually has, which lets us
+//   err on the side of sending to the backend more requests than we think it
+//   will accept for the sake of speeding up the propagation of state. A
+//   default of 2.0 is used.
+// * RequestsPadding: is used to decrease the (client-side) throttling
+//   probability in the low QPS regime (to speed up propagation of state), as
+//   well as to safeguard against hitting a client-side throttling probability
+//   of 100%. The weight of this value decreases as the number of requests in
+//   recent history grows. A default of 8 is used.
+//
+// The adaptive throttler attempts to estimate the probability that a request
+// will be throttled using recent history. Server requests (both throttled and
+// accepted) are registered with the throttler (via the RegisterBackendResponse
+// method), which then recommends client-side throttling (via the
+// ShouldThrottle method) with probability given by:
+// (requests - RatioForAccepts * accepts) / (requests + RequestsPadding)
+type Throttler struct {
+	ratioForAccepts float64
+	requestsPadding float64
+
+	// Number of total requests and throttled requests in the lookback period.
+	mu        sync.Mutex
+	requests  *lookback
+	throttled *lookback
+}
+
+// New initializes a new adaptive throttler with the default values.
+func New() *Throttler {
+	return newWithArgs(defaultDuration, defaultBins, defaultRatioForAccepts, defaultRequestsPadding)
+}
+
+// newWithArgs initializes a new adaptive throttler with the provided values.
+// Used only in unittests.
+func newWithArgs(duration time.Duration, bins int64, ratioForAccepts, requestsPadding float64) *Throttler {
+	return &Throttler{
+		ratioForAccepts: ratioForAccepts,
+		requestsPadding: requestsPadding,
+		requests:        newLookback(bins, duration),
+		throttled:       newLookback(bins, duration),
+	}
+}
+
+// ShouldThrottle returns a probabilistic estimate of whether the server would
+// throttle the next request. This should be called for every request before
+// allowing it to hit the network. If the returned value is true, the request
+// should be aborted immediately (as if it had been throttled by the server).
+func (t *Throttler) ShouldThrottle() bool {
+	randomProbability := randFunc()
+	now := timeNowFunc()
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	requests, throttled := t.requests.sum(now), t.throttled.sum(now)
+	accepts := float64(requests - throttled)
```

Comment:
Minor optimzation: if we store "accepts" as a `lookback` instead of "requests", we can do one `add()` per operation instead of 1 for accepted and 2 for throttled.

Reply:
Done.